### PR TITLE
Switch from SpeedControllers to Envelopes

### PR DESCRIPTION
**Pull request by multun**

### CODE REVIEW

File: core/src/main/java/fr/sncf/osrd/envelope/EnvelopePart.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+
+import java.util.Arrays;
+import java.util.Objects;
+
+public final class EnvelopePart {
+    public EnvelopeType type;
+    public double[] positions;
+    public double[] speeds;
+    public boolean isCoasting;
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/envelope/EnvelopePart.java]: These should probably be `final`

File: core/src/main/java/fr/sncf/osrd/envelope/EnvelopePart.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    public double getBegin() {
+        return positions[0];
+    }
+
+    public double getEnd() {
+        return positions[positions.length - 1];
+    }
+
+    public int size() {
+        assert speeds.length == positions.length;
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/envelope/EnvelopePart.java]: Is it possible that this fails? It's already checked in the constructor and I don't think we ever re-assign an array

File: core/src/main/java/fr/sncf/osrd/envelope/EnvelopePart.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        return isCoasting == that.isCoasting && type == that.type
+                && Arrays.equals(positions, that.positions) && Arrays.equals(speeds, that.speeds);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(type, isCoasting);
+        result = 31 * result + Arrays.hashCode(positions);
+        result = 31 * result + Arrays.hashCode(speeds);
+        return result;
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/envelope/EnvelopePart.java:72]: I prefer [`Objects.hash`](https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#hash-java.lang.Object...-) to merge hashes

File: core/src/main/java/fr/sncf/osrd/envelope/EnvelopeBuilder.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+        EnvelopeChangeType(int priority) {
+            this.priority = priority;
+        }
+    }
+
+    @SuppressFBWarnings(
+            value = {"EQ_COMPARETO_USE_OBJECT_EQUALS"},
+            justification = "This class has a natural ordering that is inconsistent with equals")
+    static final class EnvelopeChange implements Comparable<EnvelopeChange> {
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/envelope/EnvelopeBuilder.java]:
```suggestion
    private static final class EnvelopeChange implements Comparable<EnvelopeChange> {
```

File: core/src/main/java/fr/sncf/osrd/envelope/EnvelopeBuilder.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
+package fr.sncf.osrd.envelope;
+
+import static fr.sncf.osrd.envelope.EnvelopeBuilder.EnvelopeChangeType.*;
+
+import java.util.ArrayList;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.util.Collections;
+
+public final class EnvelopeBuilder {
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/envelope/EnvelopeBuilder.java:9]: I don't understand everything that's going on in this class, but I really like its public interface

File: core/src/main/java/fr/sncf/osrd/envelope/EnvelopePart.java (hunk 3 above)
Comment (multun) [reply core/src/main/java/fr/sncf/osrd/envelope/EnvelopePart.java:72]: this stuff is autogenerated by intellij, which didn't use objects.hash as it doesn't work for arrays

**Review by eckter — APPROVED** [review]

**Pull request merged**